File: Source Files/UNITY C#/Assets/Scripts/NEAT/Layers.cs
using Unity.Jobs;
using UnityEngine;

namespace Assets.Scripts.NEAT
{
    public class Layers
    {
        private static int BestPopulation;
        private readonly UpdateThreads updateThreads;
        public static Population[] Populations { get; private set; }

        private struct UpdateThreads : IJobParallelFor
        {
            public void Execute(int i)
            {
                if (Populations[i].AllAgentsDead())
                {
                    Populations[i].CalculateFitness();
                    Populations[i].SetBestAgent();
                    Populations[i].CalculateFitnessSum();
                    SetBestPopulation();
                    for (int j = 0; j < Populations[i].Agents.Length; j++)
                        Object.Destroy(Populations[i].Agents[j].Sphere);
                    Populations[i].NaturalSelection();
                    Populations[i].Mutate();
                    if (Populations[i].ReachedTheGoal) Populations[i].AfterReach++;
                }
                else
                    Populations[i].Update();
            }
        }

        public Layers(int directionArraySize, int populationQuantity, int layersQuantity, float mutationRate, float maxSpeed)
        {
            BestPopulation = 0;
            updateThreads = new UpdateThreads();
            Populations = new Population[layersQuantity];
            for (int i = 0; i < Populations.Length; i++)
                Populations[i] = new Population(directionArraySize, populationQuantity, layersQuantity, mutationRate, maxSpeed);
        }

        public void Update()
        {
            for (int i = 0; i < Populations.Length; i++)
                updateThreads.Execute(i);
        }

        public bool AllPopulationsDead()
        {
            for (int i = 0; i < Populations.Length; i++)
                if (!Populations[i].AllAgentsDead())
                    return false;
            return true;
        }

        public static void SetBestPopulation()
        {
            float max = 0;
            int maxIndex = 0;
            for (int i = 0; i < Populations.Length; i++)
                if (Populations[i].FitnessSum > max)
                {
                    max = Populations[i].FitnessSum;
                    maxIndex = i;
                }
            BestPopulation = maxIndex;
        }

        public Population GetBestPopulation() { return Populations[BestPopulation]; }
    };
}

File: Source Files/neat/src/Layers.cpp
//
// Created by Вадим Холошня on 17.03.2020.
//

#include "Layers.hpp"

Layers::Layers() : best_population(0)
{
    populations.resize(layers_quantity);
    threads.resize(layers_quantity);
    for (unsigned int i = 0; i < layers_quantity; i++)
        threads[i].reset(new sf::Thread(std::bind(&Layers::update_selected, this, i)));
}

bool Layers::all_populations_dead()
{
    for (auto &el : populations)
        if (!el.all_agents_dead())
            return false;
    return true;
}

void Layers::set_best_population()
{
    best_population = static_cast<int>(std::max_element(populations.begin(), populations.end(),
                                                        [](const Population &population_1,
                                                           const Population &population_2)
                                                        {
                                                            return population_1.fitness_sum <
                                                                   population_2.fitness_sum;
                                                        }) - populations.begin());
}

void Layers::update()
{
    for (auto &el : threads)
        el->launch();
    for (auto &el : threads)
        el->wait();
}

void Layers::update_selected(const int &i)
{
    if (populations[i].all_agents_dead())
    {
        populations[i].calculate_fitness();
        populations[i].set_best_agent();
        populations[i].calculate_fitness_sum();
        set_best_population();
        populations[i].natural_selection();
        populations[i].mutate();
        if (populations[i].reached_the_goal) populations[i].after_reach++;
    } else
        populations[i].update();
}

void Layers::show(sf::RenderWindow &window)
{
    for (auto &el : populations)
        el.show(window);
}
